presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSource.java

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.pagefile;
+
+import com.facebook.presto.hive.HiveColumnHandle;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.page.PagesSerde;
+import io.airlift.slice.InputStreamSliceInput;
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readPages;
+import static java.util.Objects.requireNonNull;
+
+public class PageFilePageSource
+        implements ConnectorPageSource
+{
+    private final FSDataInputStream inputStream;
+    private final Iterator<Page> pageReader;
+    private final int[] hiveColumnIndexes;
+
+    private boolean closed;
+    private long completedPositions;
+    private long completedBytes;
+    private long readTimeNanos;
+    private long memoryUsageBytes;
+
+    public PageFilePageSource(
+            FSDataInputStream inputStream,
+            PagesSerde pagesSerde,
+            List<HiveColumnHandle> columns)
+    {
+        this.inputStream = requireNonNull(inputStream, "inputStream is null");
+        pageReader = readPages(
+                requireNonNull(pagesSerde, "pagesSerdeis null"),
+                new InputStreamSliceInput(inputStream));
+
+        int size = requireNonNull(columns, "columns is null").size();
+        this.hiveColumnIndexes = new int[size];
+
+        for (int columnIndex = 0; columnIndex < size; columnIndex++) {
+            HiveColumnHandle column = columns.get(columnIndex);
+            hiveColumnIndexes[columnIndex] = column.getHiveColumnIndex();
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return completedPositions;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return readTimeNanos;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed || !pageReader.hasNext();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (isFinished()) {
+            return null;
+        }
+        long start = System.nanoTime();
+
+        Page page = pageReader.next();
+
+        Block[] blocks = new Block[hiveColumnIndexes.length];
+        for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
+            if (hiveColumnIndexes[fieldId] >= page.getChannelCount()) {
+                throw new PrestoException(
+                        NOT_SUPPORTED,
+                        "schema mismatched table is not supported for PageFile format");
```

Comment:
We call it `schema evolution`. Let's change it to `schema evolution is not supported for page file format`